rts/javascript/values.js

```diff
@@ -0,0 +1,205 @@
+                  "BigInt64Array" : ' i64',
+                  "Uint8Array" : '  u8',
+                  "Uint16Array" :  ' u16',
+                  "Uint32Array" :  ' u32',
+                  "BigUint64Array" :  ' u64',
+                  "Float32Array" : ' f32',
+                  "Float64Array" : ' f64',
+                };
+
+var typToType = { '  i8' : Int8Array ,
+                  ' i16' : Int16Array ,
+                  ' i32' : Int32Array ,
+                  ' i64' : BigInt64Array ,
+                  '  u8' : Uint8Array ,
+                  ' u16' :  Uint16Array ,
+                  ' u32' :  Uint32Array ,
+                  ' u64' :  BigUint64Array ,
+                  ' f32' : Float32Array ,
+                  ' f64' : Float64Array ,
+                  'bool' : Uint8Array
+                };
+
+function binToStringArray(buff, array) {
+  for (var i = 0; i < array.length; i++) {
+    array[i] = buff[i];
+  }
+}
+
+function fileToBuff(fname) {
+  var readline = require('readline');
+  var fs = require('fs');
+  var buff =  fs.readFileSync(fname);
+  return buff;
+}
+
+function fileToValue(fname) {
+  var str = fileToBuff(fname);
+  return read_binary(str);
+}
+
+function read_binary(buff) {
+  // Skip leading white space
+  while (buff.slice(0, 1).toString().trim() == "") {
+    buff = this.buff.slice(1);
+  }
+  if (buff[0] != 'b'.charCodeAt(0)) {
+    throw "Not in binary format"
+  }
+  var version = buff[1];
+  if (version != 2) {
+    throw "Not version 2";
+  }
+  var num_dim = buff[2];
+  var typ = buff.slice(3, 7);
+  buff = buff.slice(7);
+  if (num_dim == 0) {
+    return read_bin_scalar(buff, typ);
+  } else {
+    return read_bin_array(buff, num_dim, typ);
+  }
+}
+
+var typToSize = {
+  "bool" : 1,
+  "  u8" : 1,
+  "  i8" : 1,
+  " u16" : 2,
+  " i16" : 2,
+  " u32" : 4,
+  " i32" : 4,
+  " f32" : 4,
+  " u64" : 8,
+  " i64" : 8,
+  " f64" : 8,
+}
+
+function toU8(ta) {
+  return new Uint8Array(ta.buffer, ta.byteOffset, ta.byteLength);
+}
+
+
+function construct_binary_value(v, typ) {
+  var dims;
+  var payload_bytes;
+  var filler;
+  if (v instanceof FutharkOpaque) {
+    throw "Opaques are not supported";
+  } else if (v instanceof FutharkArray) {
+    var t = v.futharkType();
+    var ftype = "    ".slice(t.length) + t;
+    var shape = v.shape();
+    var ta = v.toTypedArray(shape);
+    var da = new BigInt64Array(shape);
+    dims = shape.length;
+    payload_bytes = da.byteLength + ta.byteLength;
+    filler = (bytes) => {
+      bytes.set(toU8(da), 7);
+      bytes.set(toU8(ta), 7 + da.byteLength);
+    }
+  } else {
+    var ftype = "    ".slice(typ.length) + typ;
+    dims = 0;
+    payload_bytes = typToSize[ftype];
+    filler = (bytes) => {
+      var scalar = new (typToType[ftype])([v]);
+      bytes.set(toU8(scalar), 7);
+    }
+  }
+  var total_bytes = 7 + payload_bytes;
+  var bytes = new Uint8Array(total_bytes);
+  bytes[0] = Buffer.from('b').readUInt8();
+  bytes[1] = 2; // Not sure why this
```

Comment:
This is the version number of the binary data format (just in case we ever change it again, which we did do many years ago).